add testing sdk docs (#70)
cloudquery/docs: docs/developers/sdk/testing.md

# Testing

> Testing providers is key factor in ensuring the quality of your provider and developer experience, so future contributors can easily test their changes against.

Reviewer: against... ? Smells like an incomplete sentence, I think

> Testing your provider is key in ensuring quality of your provider and developer experiance for future contributors that can easily test their changes against the tests.

Reviewer 2: ```suggestion
Testing providers is key factor in ensuring the quality of your provider and developer experience, so future contributors can easily test their changes against.
```

> CloudQuery SDK provides utility test function to help test each resource and verify it's output.

Reviewer 2: ```suggestion
CloudQuery SDK provides an utility test function to help you test each resource and verify it's output.
```

Author: an utility ? a utility maybe ?

## Requirements

> The testing framework provided by CloudQuery SDK is essentially an end-to-end (integration) tests. This means you will need a real test account and real resources in place. It is suggested to create those test resource with any IaC tool (Terraform, Pulumi, etc...).

Reviewer 2: ```suggestion
The testing framework provided by CloudQuery SDK are essentially an end-to-end (integration) tests. This means you will need a real test account and real resources in place. It is suggested to create those test resources with any IaC tool (Terraform, Pulumi, etc...).
```
either `are ... e2e tests` or is ... e2e test`

> The testing framework provided by CloudQuery SDK are essentially an end-to-end (integration) tests. This means you will need a real test account and real resources in place. It is suggested to create those test resources with any IaC tool (Terraform, Pulumi, etc...).

Reviewer: ```suggestion
The testing framework provided by CloudQuery SDK are essentially end-to-end (integration) tests. This means you will need a real test account and real resources in place. It is suggested to create those test resources with any IaC tool (Terraform, Pulumi, etc...).
```

## How Resource Tests work

There are two main goals to CloudQuery testing framework:

- Check that the extract(fetch) functioanality is defined correctly and works without errors against real infrastructures.

> - Check the schema definition and all columns resolvers that are defined correctly and the data exists after a successful fetch.

Reviewer: ```suggestion
- Check the schema definition and all column resolvers are defined correctly, the data exists after a successful fetch.
```

> Here is the flow of each test:

Reviewer 2: Maybe  subtitle it?

```suggestion
### Test flow execution:
```

> - The resource table and all it's children tables are delted from PostgreSQL to work with clean data.

Reviewer: ```suggestion
- The resource table and all it's children tables are deleted from PostgreSQL to work with a clean state.
```

> - CloudQuery SDK is fetching the resource, transforming and loading it into PostgreSQL.

Reviewer: ```suggestion
- CloudQuery SDK fetches the resource, transforms and loads it into PostgreSQL.
```

> - For each resource table (and it's children) CloudQuery is reading all the rows and checking that 

Reviewer: ```suggestion
- For each resource table (and it's children) CloudQuery reads all the rows and checks that 
```

    - Each table has at least one row available unless `IgnoreInTests` is defined in the `Table`.
    - Each column has data which means the default or custom column resolver worked (unless skipped with `IgnoreInTests` on the `Column`).

## ResourceTest and ResourceTestCase

> Each resource can be tested end-to-end with `ResourceTest` function. Here is an example

Reviewer: ```suggestion
Each resource can be tested end-to-end with `ResourceTest` function. Here is an example:
```
